File: scraper.js
import puppeteer from 'puppeteer';
import fs from 'fs';
import path from 'path';

const essentialsUrl = 'https://knowledge.workspace.google.com/admin/getting-started/editions/compare-essentials-editions';
const frontlineUrl = 'https://knowledge.workspace.google.com/admin/getting-started/editions/compare-frontline-editions?sjid=17864279373018464028-NC&visit_id=639104993134306662-4217303682&rd=1#core';

async function scrapeTables(url) {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.goto(url, { waitUntil: 'networkidle2' });

  // Expanding any "expand all" buttons if they exist
  try {
    const expandButtons = await page.$$('button');
    for (const btn of expandButtons) {
      const text = await page.evaluate(el => el.textContent, btn);
      if (text && text.toLowerCase().includes('expand all')) {
        await btn.click();
      }
    }
    // Wait a bit for animations
    await new Promise(r => setTimeout(r, 1000));
  } catch (e) {
    console.log('No expand all buttons found or failed to click');
  }

  const data = await page.evaluate(() => {
    const result = [];
    const tables = document.querySelectorAll('table');
    
    tables.forEach(table => {
      const headers = [];
      const thead = table.querySelector('thead');
      if (thead) {
        thead.querySelectorAll('tr').forEach(tr => {
          const rowHeaders = Array.from(tr.querySelectorAll('th, td')).map(cell => cell.innerText.trim());
          if (rowHeaders.length > 0) headers.push(rowHeaders);
        });
      } else {
        const firstTr = table.querySelector('tbody tr');
        if (firstTr) {
          headers.push(Array.from(firstTr.querySelectorAll('th, td')).map(cell => cell.innerText.trim()));
        }
      }

      const rows = Array.from(table.querySelectorAll('tbody tr'));
      const tableData = [];
      rows.forEach(row => {
        const cells = Array.from(row.querySelectorAll('td, th')).map(td => {
            // Check for checkmarks or icons
            const icon = td.querySelector('span.material-icons, span.google-material-icons');
            if (icon && icon.innerText.trim() === 'done') {
                return 'Included';
            } else if (icon && icon.innerText.trim() === 'close') {
                return 'Not included';
            }
            return td.innerText.trim();
        });
        if (cells.length > 0) {
          tableData.push(cells);
        }
      });
      
      result.push({
        headers,
        rows: tableData
      });
    });
    return result;
  });

  await browser.close();
  return data;
}

async function main() {
  console.log('Scraping Essentials...');
  const essentialsData = await scrapeTables(essentialsUrl);
  
  console.log('Scraping Frontline...');
  const frontlineData = await scrapeTables(frontlineUrl);
  
  const dataDir = path.join(process.cwd(), 'src', 'data');
  if (!fs.existsSync(dataDir)) {
    fs.mkdirSync(dataDir, { recursive: true });
  }

  fs.writeFileSync(path.join(dataDir, 'raw_essentials.json'), JSON.stringify(essentialsData, null, 2));
  fs.writeFileSync(path.join(dataDir, 'raw_frontline.json'), JSON.stringify(frontlineData, null, 2));
  console.log('Scraping completed and saved to src/data/raw_*.json');
}

main();

File: scraper.py
import urllib.request
import json
from html.parser import HTMLParser
import re

class TableParser(HTMLParser):
    def __init__(self):
        super().__init__()
        self.in_table = False
        self.in_row = False
        self.in_cell = False
        self.in_header = False
        
        self.tables = []
        self.current_table = []
        self.current_row = []
        self.current_cell = []
        
        self.current_header_text = []
        self.last_header = "General"

    def handle_starttag(self, tag, attrs):
        if tag in ('h2', 'h3'):
            self.in_header = True
            self.current_header_text = []
        elif tag == 'table':
            self.in_table = True
            self.current_table = []
        elif tag == 'tr' and self.in_table:
            self.in_row = True
            self.current_row = []
        elif tag in ('td', 'th') and self.in_row:
            self.in_cell = True
            self.current_cell = []

    def handle_data(self, data):
        if self.in_header:
            text = data.strip()
            if text:
                self.current_header_text.append(text)
        elif self.in_cell:
            text = data.strip()
            if text:
                self.current_cell.append(text)

    def handle_endtag(self, tag):
        if tag in ('h2', 'h3'):
            self.in_header = False
            if self.current_header_text:
                self.last_header = ' '.join(self.current_header_text).strip()
        elif tag == 'table':
            self.in_table = False
            if self.current_table:
                valid_rows = [r for r in self.current_table if len(r) > 1]
                if valid_rows:
                    self.tables.append({
                        "section": self.last_header,
                        "rows": valid_rows
                    })
        elif tag == 'tr' and self.in_table:
            self.in_row = False
            if self.current_row:
                self.current_table.append(self.current_row)
        elif tag in ('td', 'th') and self.in_row:
            self.in_cell = False
            cell_text = ' '.join(self.current_cell).replace("  ", " ").strip()
            if 'done' in cell_text.split():
                cell_text = 'Included'
            elif 'close' in cell_text.split() and len(cell_text.split()) == 1:
                cell_text = 'Not included'
            self.current_row.append(cell_text)

def fetch_tables(url):
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    with urllib.request.urlopen(req) as response:
        html = response.read().decode('utf-8')
    parser = TableParser()
    parser.feed(html)
    return parser.tables

essentials_url = "https://knowledge.workspace.google.com/admin/getting-started/editions/compare-essentials-editions"
frontline_url = "https://knowledge.workspace.google.com/admin/getting-started/editions/compare-frontline-editions"

print("Fetching Essentials...")
e_tables = fetch_tables(essentials_url)
print("Fetching Frontline...")
f_tables = fetch_tables(frontline_url)

with open('data.json', 'w') as f:
    json.dump({"essentials": e_tables, "frontline": f_tables}, f, indent=2)
print(f"Done. Wrote data.json with {len(e_tables)} essentials tables and {len(f_tables)} frontline tables")
